### PR TITLE
Fail with error when autodiscover providers have no defined configs

### PR DESCRIPTION
**Pull request by andrewvc**
If one accidentally misconfigures an autodiscover provider and there are no defined configs no error is raised today. This patch adds an error in this situation and also does a better job adding context to other initialization errors.

You can test this with the following config which looks like it should work in heartbeat:

```
heartbeat.autodiscover:
  providers:
    - type: docker
      templates:
          condition:
            contains:
              docker.container.image: redis
          config:
            - type: tcp
              hosts: ["${data.host}:${data.port}"]
              schedule: "@every 1s"
              timeout: 1s

```

However, it does not because `templates` should be an array. The correct config is below:

```
heartbeat.autodiscover:
  providers:
    - type: docker
      templates:
        - condition:
            contains:
              docker.container.image: redis
          config:
            - type: tcp
              hosts: ["${data.host}:${data.port}"]
              schedule: "@every 1s"
              timeout: 1s
```

### CODE REVIEW

File: libbeat/autodiscover/template/config.go [hunk 1]
```diff
@@ -43,6 +45,10 @@ type MapperSettings []*struct {
 
 // NewConfigMapper builds a template Mapper from given settings
 func NewConfigMapper(configs MapperSettings) (mapper Mapper, err error) {
+	if len(configs) == 0 {
```
Comment (exekias) [inline libbeat/autodiscover/template/config.go]: This is a valid config, for instance when people is using hints based autodiscover, it's ok to leave templates empty:

```
filebeat.autodiscover:
  providers:
    - type: kubernetes
      hints.enabled: true
```

We would need to account for that for an error like this

File: libbeat/autodiscover/template/config.go [hunk 2]
```diff
@@ -43,6 +45,10 @@ type MapperSettings []*struct {
...
+		return nil, fmt.Errorf("no configs defined for autodiscover provider")
+	}
+
 	for _, c := range configs {
 		condMap := &ConditionMap{Configs: c.Configs}
```
Comment (exekias) [inline libbeat/autodiscover/template/config.go]: Another check we could do here is to ensure `len(c.Configs) > 0`. I think in most cases this is were bad configs happen? when the user does:

```
autodiscover:
  providers:
    type: kubernetes
    templates:
      - equals:
          container.image: ...
        config:
          type: container
```

But config is expected to be a list, so a `-` is missing

Comment (andrewvc) [reply libbeat/autodiscover/template/config.go]: I'm a bit confused, is this section used when hints are enabled?

**Comment by andrewvc** [comment]
@exekias I just pushed a fix that takes hints into account for providers that support them.

**Comment by exekias** [comment]
sorry, I approved as the change looks good to me, but it seems one test is failing, we will probably need to adjust it after this change. A changelog will also be needed

**Comment by Randy-312** [comment]
Kudos on this one!... it would've saved us a lot of time when we had our configuration mis-done.

**Comment by andrewvc** [comment]
Sorry for the delay here, hoping to come back to this one within the next few weeks, I'm sidetracked on other PRs ATM

**Pull request merged**